### DISCUSSION TITLE
How to use with NetworkExtension framework

### DISCUSSION THREAD

**Discussion by waqqasjabbar** [iterate-ch/rococoa#48] [Q&A]
Hi
I want to call functions in the NetworkExtension Framework. Any ideas on how to do that would be helpful. Specifically, I want to call [`loadAllFromPreferencesWithCompletionHandler`](https://developer.apple.com/documentation/networkextension/netunnelprovidermanager/1406271-loadallfrompreferenceswithcomple) within `NETunnelProviderManager`

I am giving it a [try (code here)](https://github.com/safekids-ai/packet-proxy-java/tree/rococoa) but getting a VM crash. Any help would be appreciated.

Regards
Waqqas

**Comment by dkocher** [comment]
Completion handler aka callbacks are not supported. You will have to write a proxy ObjC class with a function like (pseudo code in Swift)

```
public class CallbackNETunnelProviderManager: NSObject {
@objc public class func loadAllFromPreferences(completionHandler: MyCompletionHandlerCallback
{
        NETunnelProviderManager. loadAllFromPreferences(completionHandler: { error in
            completionHandler?.didFinish(nil, withError: error)
        })
}

@objc(FileProviderManagerConnectCallback)
public protocol MyCompletionHandlerCallback {
    func didFinish(_: AnyObject?, withError error: Error?)
}

```
